### PR TITLE
don't gracefully delete terminated pods

### PR DESCRIPTION
**Pull request by mikedanese**

### CODE REVIEW

**Comment by smarterclayton** [comment]
LGTM

**Comment by smarterclayton** [comment]
Add a test please though

File: pkg/registry/pod/strategy.go [hunk 1]
```diff
@@ -109,6 +109,10 @@ func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOp
 	if len(pod.Spec.NodeName) == 0 {
 		period = 0
 	}
+	// if the pod is already terminated, delete immediately
+	if pod.Status.Phase == api.PodFailed && pod.Status.Phase == api.PodSucceeded {
```
Comment (bgrant0607) [inline pkg/registry/pod/strategy.go]: OR, not AND

**Comment by mikedanese** [comment]
Added test for defaulted DeleteOptions

File: pkg/registry/pod/strategy_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import (
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+)
+
+func TestCheckGracefulDelete(t *testing.T) {
+
+	defaultGracePeriod := int64(30)
+	otherGracePeriod := int64(15)
```
Comment (ncdc) [inline pkg/registry/pod/strategy_test.go]: unused?

File: pkg/registry/pod/strategy_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			in: &api.Pod{
+				Spec:   api.PodSpec{NodeName: "something"},
+				Status: api.PodStatus{Phase: api.PodFailed},
+			},
+			gracePeriod: 0,
+		},
+		{
+			in: &api.Pod{
+				Spec:   api.PodSpec{},
+				Status: api.PodStatus{Phase: api.PodPending},
```
Comment (ncdc) [inline pkg/registry/pod/strategy_test.go]: Should this be PodSucceeded?

Comment (mikedanese) [inline pkg/registry/pod/strategy_test.go]: Testing that we don't gracefully terminate pods without NodeName assigned, even if they are non-terminated, e.g. before they are scheduled.

File: pkg/registry/pod/strategy_test.go (hunk 2 above)
Comment (mikedanese) [inline pkg/registry/pod/strategy_test.go]: fixed

**Comment by smarterclayton** [comment]
ok to test

**Comment by mikedanese** [comment]
Flake Kubernetes e2e suite.Probing container with readiness probe should not be ready before initial delay and never restart [Conformance]

@k8s-bot ok to test

File: pkg/registry/pod/strategy_test.go (hunk 3 above)
Comment (ncdc) [inline pkg/registry/pod/strategy_test.go]: What I really meant to write was "do you want to have a test case that checks that gracePeriod is set to 0 when the phase is succeeded?" I looked at the change to strategy.go and it seems like that would be a useful test based on the code change.

Comment (mikedanese) [inline pkg/registry/pod/strategy_test.go]: It seemed redundant to the PodFailed test case but will add

**Comment by mikedanese** [comment]
@ncdc PTAL

**Comment by ncdc** [comment]
@mikedanese looks good, thanks!

**Comment by ihmccreery** [comment]
@mikedanese Can you please file an issue under `kind/workaround-removal` for removing this when we drop support for `v1.0`?  Thanks!

**Comment by mikedanese** [comment]
@ihmccreery we want this forever.

**Pull request merged**